src/constraints/RebarBondSlipConstraint.C

```diff
@@ -0,0 +1,517 @@
+  if (_output_axial_force)
+    _output_axial_force->setNodalValue(MetaPhysicL::raw_value(_bond_stress));
+  if (_output_axial_plastic_slip)
+    _output_axial_plastic_slip->setNodalValue(plastic_slip);
+}
+
+template <bool is_ad>
+GenericReal<is_ad>
+RebarBondSlipConstraintTempl<is_ad>::computeQpResidual(Moose::ConstraintType type)
+{
+  GenericReal<is_ad> resid = _constraint_residual(_component);
+  switch (type)
+  {
+    case Moose::Secondary:
+      return resid * _test_secondary[_i][_qp];
+
+    case Moose::Primary:
+      return -resid * _test_primary[_i][_qp];
+  }
+
+  return 0.0;
+}
+
+template <>
+Real
+RebarBondSlipConstraintTempl<true>::computeQpJacobian(Moose::ConstraintJacobianType /*type*/)
+{
+  mooseError("In ADRebarBondSlipConstraint, computeQpJacobian should not be called. Check "
+             "computeQpJacobian "
+             "implementation.");
+  return 0.0;
+}
+
+template <>
+Real
+RebarBondSlipConstraintTempl<false>::computeQpJacobian(Moose::ConstraintJacobianType type)
+{
+
+  Real jac_axial = _constraint_jacobian_axial(_component);
+
+  switch (type)
+  {
+    case Moose::SecondarySecondary:
+      return _phi_secondary[_j][_qp] * jac_axial * _secondary_node_tangent(_component) *
+                 _test_secondary[_i][_qp] +
+             _phi_secondary[_j][_qp] * this->_penalty * _test_secondary[_i][_qp] *
+                 (1.0 - _secondary_node_tangent(_component) * _secondary_node_tangent(_component));
+
+    case Moose::SecondaryPrimary:
+      return -_phi_primary[_j][_qp] * jac_axial * _secondary_node_tangent(_component) *
+                 _test_secondary[_i][_qp] -
+             _phi_primary[_j][_qp] * this->_penalty * _test_secondary[_i][_qp] *
+                 (1.0 - _secondary_node_tangent(_component) * _secondary_node_tangent(_component));
+
+    case Moose::PrimarySecondary:
+      return -_test_primary[_i][_qp] * jac_axial * _secondary_node_tangent(_component) *
+                 _phi_secondary[_j][_qp] -
+             _test_primary[_i][_qp] * this->_penalty * _phi_secondary[_j][_qp] *
+                 (1.0 - _secondary_node_tangent(_component) * _secondary_node_tangent(_component));
+
+    case Moose::PrimaryPrimary:
+      return _test_primary[_i][_qp] * jac_axial * _secondary_node_tangent(_component) *
+                 _phi_primary[_j][_qp] +
+             _test_primary[_i][_qp] * this->_penalty * _phi_primary[_j][_qp] *
+                 (1.0 - _secondary_node_tangent(_component) * _secondary_node_tangent(_component));
+
+    default:
+      mooseError("Unsupported type");
+      break;
+  }
+  return 0.0;
+}
+
+template <>
+Real
+RebarBondSlipConstraintTempl<true>::computeQpOffDiagJacobian(Moose::ConstraintJacobianType /*type*/,
+                                                             unsigned int /*jvar*/)
+{
+  mooseError("In ADRebarBondSlipConstraint, computeQpOffDiagJacobian should not be called. Check "
+             "computeQpJacobian implementation.");
+  return 0.0;
```

Comment:
I think you could get fancy and put `[[noreturn]]` in the declaration in the header file to avoid returning.

Reply:
I don't think I can do that here because there is an implementation for RebarBondSlipConstraintTempl<false>::computeQpOffDiagJacobian.